lib/pact/api/exec_command_request.ex

```diff
@@ -0,0 +1,262 @@
+defmodule Kadena.Pact.API.ExecCommandRequest do
+  @moduledoc """
+  Specifies functions to build PACT execution command requests.
+  """
+
+  alias Kadena.Chainweb.Pact.JSONPayload
+  alias Kadena.Cryptography.Sign
+  alias Kadena.Pact.API.CommandRequest
+
+  alias Kadena.Types.{
+    Command,
+    CommandPayload,
+    EnvData,
+    ExecPayload,
+    KeyPair,
+    MetaData,
+    NetworkID,
+    PactPayload,
+    SignaturesList,
+    SignCommand,
+    Signer,
+    SignersList
+  }
+
+  @behaviour CommandRequest
+
+  @type cmd :: String.t()
+  @type code :: String.t()
+  @type command :: Command.t()
+  @type data :: EnvData.t() | nil
+  @type hash :: String.t()
+  @type keypair :: KeyPair.t()
+  @type keypairs :: list(keypair())
+  @type json_string_payload :: String.t()
+  @type network_id :: NetworkID.t()
+  @type nonce :: String.t()
+  @type meta_data :: MetaData.t()
+  @type pact_payload :: PactPayload.t()
+  @type signer :: Signer.t()
+  @type signers :: SignersList.t()
+  @type signatures :: SignaturesList.t()
+  @type sign_command :: SignCommand.t()
+  @type sign_commands :: list(sign_command())
+  @type valid_command :: {:ok, command()}
+  @type valid_command_json_string :: {:ok, json_string_payload()}
+  @type valid_hash :: {:ok, hash()}
+  @type valid_payload :: {:ok, pact_payload()}
+  @type valid_sign_command :: {:ok, [sign_command()]}
+
+  @type t :: %__MODULE__{
+          network_id: network_id(),
+          code: code(),
+          data: data(),
+          nonce: nonce(),
+          meta_data: meta_data(),
+          keypairs: keypairs(),
+          signers: signers()
+        }
+
+  defstruct [:network_id, :meta_data, :code, :nonce, :signers, :data, keypairs: []]
+
+  @impl true
+  def new(opts \\ nil)
+  def new(nil), do: %__MODULE__{}
+
+  def new(opts) when is_list(opts) do
+    network_id = Keyword.get(opts, :network_id)
+    code = Keyword.get(opts, :code, "")
+    data = Keyword.get(opts, :data)
+    nonce = Keyword.get(opts, :nonce, "")
+    meta_data = Keyword.get(opts, :meta_data, %MetaData{})
+    keypairs = Keyword.get(opts, :keypairs, [])
+    signers = Keyword.get(opts, :signers, %SignersList{})
+
+    %__MODULE__{}
+    |> set_network(network_id)
+    |> set_data(data)
+    |> set_code(code)
+    |> set_nonce(nonce)
+    |> set_metadata(meta_data)
+    |> add_keypairs(keypairs)
+    |> add_signers(signers)
+  end
+
+  def new(_opts), do: {:error, [args: :not_a_list]}
+
+  @impl true
+  def set_network(%__MODULE__{} = cmd_request, network) do
+    case NetworkID.new(network) do
+      %NetworkID{} = network_id -> %{cmd_request | network_id: network_id}
+      {:error, reason} -> {:error, [network_id: :invalid] ++ reason}
+    end
+  end
+
+  def set_network({:error, reason}, _network), do: {:error, reason}
+
+  @impl true
+  def set_data(%__MODULE__{} = cmd_request, nil), do: %{cmd_request | data: nil}
```

Comment:
the data default value is `nil`, so why it allows `nil` to be added again? I would remove this line.